logagg_collector/collector.py

```diff
@@ -0,0 +1,926 @@
+
+            else:
+                err_msg = get_topic_info_result['result'].get('details')
+                raise Exception(err_msg)
+
+
+    def _init_logaggfs_paths(self, logaggfs_dir):
+        '''
+        Logaggfs directories and file initialization
+
+        >>> test_dir = utils.ensure_dir('/tmp/xyz')
+        >>> trackfile = open(test_dir+'/trackfiles.txt', 'w+'); trackfile.close()
+        >>> lc = collector = LogCollector('localhost', '1088', None, test_dir, test_dir)
+
+        >>> lc._init_logaggfs_paths(test_dir)
+        AttrDict({'logcache': '/tmp/xyz', 'logs_dir': '/tmp/xyz/logs', 'trackfiles': '/tmp/xyz/trackfiles.txt'})
+
+        >>> import shutil; shutil.rmtree(test_dir)
+        '''
+        logaggfs = AttrDict()
+        logaggfs.logcache = logaggfs_dir
+        logaggfs.logs_dir = os.path.abspath(os.path.join(logaggfs.logcache, 'logs'))
+        logaggfs.trackfiles = os.path.abspath(os.path.join(logaggfs.logcache, 'trackfiles.txt'))
+        return logaggfs
+
+
+    def _ensure_trackfiles_sync(self):
+        '''
+        Make sure fpaths in logagg state-file are present in logaggfs trackfiles on start up
+
+        >>> import tempfile
+        >>> temp_dir = tempfile.TemporaryDirectory()
+        >>> temp_dir_path = temp_dir.name
+        >>> trackfile = open(temp_dir_path+'/trackfiles.txt', 'w+'); trackfile.close()
+        >>> lc = collector = LogCollector('localhost', '1088', None, temp_dir_path, temp_dir_path)
+
+        >>> with open(lc.logaggfs.trackfiles) as f: f.read()
+        '/var/log/serverstats/serverstats.log\\n'
+        >>> lc.state['fpaths'] = [{'fpath': '/var/log/some.log'}]; lc.state.flush()
+        >>> lc._ensure_trackfiles_sync()
+        >>> with open(lc.logaggfs.trackfiles) as f: f.read()
+        '/var/log/serverstats/serverstats.log\\n/var/log/some.log\\n'
+
+        >>> temp_dir.cleanup()
+        '''
+        # If all the files are present in trackfiles
+        for f in self.state['fpaths']:
+            if not self._fpath_in_trackfiles(f['fpath']):
+                self.add_to_logaggfs_trackfile(f['fpath'])
+
+    def _remove_redundancy(self, log):
+        '''
+        Removes duplicate data from 'data' inside log dictionary and brings it out
+
+        >>> import tempfile
+        >>> temp_dir = tempfile.TemporaryDirectory()
+        >>> temp_dir_path = temp_dir.name
+        >>> trackfile = open(temp_dir_path+'/trackfiles.txt', 'w+'); trackfile.close()
+        >>> lc = collector = LogCollector('localhost', '1088', None, temp_dir_path, temp_dir_path)
+
+        >>> log = {'id' : 46846876, 'type' : 'log',
+        ...         'data' : {'a' : 1, 'b' : 2, 'type' : 'metric'}}
+        >>> from pprint import pprint
+        >>> pprint(lc._remove_redundancy(log))
+        {'data': {'a': 1, 'b': 2}, 'id': 46846876, 'type': 'metric'}
+
+        >>> temp_dir.cleanup()
+        '''
+        for key in log:
+            if key in log and key in log['data']:
```

Comment:
I dont think we need to check
`if key in log` because we are iterating on log itself key will definitely be there in log.